src/dataset-operator/pkg/controller/dataset/dataset_controller.go

```diff
@@ -105,132 +127,295 @@ func (r *ReconcileDataset) Reconcile(request reconcile.Request) (reconcile.Resul
+			// to re-check the status
+			return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+
+		case v1.ClaimBound:
+			// if our PVC is bound make sure that the Provision Status of Dataset is
+			// shown as OK and requeue
+			if datasetInstance.Status.Provision.Status != comv1alpha1.StatusOK {
+				datasetInstance.Status.Provision.Status = comv1alpha1.StatusOK
+				datasetInstance.Status.Provision.Info = ""
+				return updateDatasetStatus(r.client, datasetInstance, reqLogger)
+			}
 		}
 	}
 
+	// everything OK don't requeue
+	// TODO maybe requeue with some delay if we
+	// need to monitor statuses
 	return reconcile.Result{}, nil
 }
 
+// initializeDataset initializing the Status of Dataset and
+// returns true if it modified the Status of the Dataset or
+// false otherwise
+func initializeDatasetStatus(d *comv1alpha1.Dataset) bool {
+	modifiedDatasetStatus := false
+	if d.Status.Provision.Status == comv1alpha1.StatusEmpty {
+		d.Status.Provision.Status = comv1alpha1.StatusInitial
+		modifiedDatasetStatus = true
+	}
+	if d.Status.Caching.Status == comv1alpha1.StatusEmpty {
+		d.Status.Caching.Status = comv1alpha1.StatusInitial
+		modifiedDatasetStatus = true
+	}
+	return modifiedDatasetStatus
+}
+
+func updateDatasetStatus(c client.Client, d *comv1alpha1.Dataset, logger logr.Logger) (reconcile.Result, error) {
+	err := c.Status().Update(context.TODO(), d)
+	if err != nil {
+		logger.Error(err, "Error updating dataset status")
+	}
+	return reconcile.Result{}, err
+}
+
+func updateDataset(c client.Client, d *comv1alpha1.Dataset, logger logr.Logger) (reconcile.Result, error) {
+	err := c.Update(context.TODO(), d)
+	if err != nil {
+		logger.Error(err, "Error updating dataset status")
+	}
+	return reconcile.Result{}, err
+}
+
+func readEventsForPVC(c *kubernetes.Clientset, pvc *v1.PersistentVolumeClaim, logger logr.Logger) (string, error) {
+	lastMessage := ""
+
+	listOpts := metav1.ListOptions{
+		FieldSelector: fmt.Sprintf(
+			"involvedObject.name=%s,"+
+				"involvedObject.resourceVersion=%s,"+
+				"involvedObject.kind=PersistentVolumeClaim,"+
+				"reason=ProvisioningFailed",
+			pvc.Name, pvc.ResourceVersion,
+		),
+	}
+	eventList, err := c.CoreV1().Events(pvc.Namespace).List(listOpts)
+	if err == nil {
+		eventsLen := len(eventList.Items)
+		if eventsLen > 0 {
+			// Assuming that always the last event in the list will be
+			// the latest?!
+			lastMessage = eventList.Items[eventsLen-1].Message
+		}
+	} else {
+		logger.Error(err, "Reading events failed")
+	}
+	return lastMessage, err
+}
+
+func datasetHasCachingAnnotationsSet(d *comv1alpha1.Dataset) (string, bool) {
+	cachingPluginName := ""
+	datasetHasCachingAnnotations := true
+	for _, cachingLabel := range []string{"dlf-plugin-type", "dlf-plugin-name"} {
+		value, exists := d.Annotations[cachingLabel]
+		if !exists {
+			datasetHasCachingAnnotations = false
+			break
+		}
+		cachingPluginName = value
+	}
+	return cachingPluginName, datasetHasCachingAnnotations
+}
+
```

Comment:
As an optimization I would take the cachingPluginName from the annotations directly if datasetHasCachingAnnotations is true

Reply:
Thanks for this amazing catch! Fixed in https://github.com/IBM/dataset-lifecycle-framework/pull/77/commits/79f7cfbc38d15cc520c3f21bdf37c6005a0acd71